Feature: File-scoped 'using' declarations for module imports

## Feature Description

Add support for file-scoped `using` declarations to make imported modules available throughout an entire file, eliminating the need to repeat `using` statements in every function.

## Motivation

Currently, `using` must be declared inside each function, leading to repetition:

```ez
import @std
import @arrays

do main() {
    using std      // Required in every function
    using arrays   // Required in every function
    println("Hello")
    push(myArray, 5)
}

do helper() {
    using std      // Must repeat
    using arrays   // Must repeat
    println("World")
}
```

With file-scoped `using`, this becomes:

```ez
import @std
import @arrays

using std      // Available in entire file
using arrays   // Available in entire file

do main() {
    println("Hello")    // std available without redeclaring
    push(myArray, 5)    // arrays available without redeclaring
}

do helper() {
    println("World")    // std still available
}
```

## Proposed Syntax

### Basic File-Scoped Usage

```ez
// 1. Imports first
import @std
import @arrays
import @strings

// 2. File-scoped 'using' declarations (must come after imports)
using std
using arrays

// 3. Other declarations
const Person struct {
    name string
    age int
}

// 4. Functions
do main() {
    // std and arrays are available here
    println("Hello World")
    push(myArray, 42)
}

do helper() {
    // std and arrays are still available here
    println("Helper function")
}
```

### Comma-Separated Syntax (Programmer's Choice)

Both styles are valid and can be mixed:

```ez
import @std
import @arrays
import @strings

// Style 1: One per line
using std
using arrays
using strings

// Style 2: Comma-separated
using std, arrays, strings

// Style 3: Mixed (programmer's choice)
using std, arrays
using strings
```

### Function-Scoped `using` ADDS to File-Scoped

Function-scoped `using` declarations add to (not replace) file-scoped modules:

```ez
import @std
import @arrays
import @strings

using std      // File-scoped
using arrays   // File-scoped

do main() {
    // Have: std, arrays
    println("Hello")    // std.println
    push(arr, 1)        // arrays.push
    
    using strings       // Add strings for this function only
    concat("a", "b")    // strings.concat
}

do helper() {
    // Have: std, arrays (NOT strings)
    println("World")
}
```

## Design Rules

### 1. Placement Rules

File-scoped `using` declarations **must**:
- ✅ Come after ALL `import` statements
- ✅ Come before any other declarations (structs, consts, functions, etc.)

```ez
// ✅ CORRECT ORDER
import @std
import @arrays
using std
using arrays
const MyStruct struct { }
do main() { }

// ❌ WRONG: using before imports
using std
import @std

// ❌ WRONG: using after declarations
import @std
const MyStruct struct { }
using std  // Too late!
```

### 2. Scope Interaction

- **File-scoped `using`**: Available in all functions in that file
- **Function-scoped `using`**: Adds modules to that specific function only
- Function-scoped declarations ADD to file-scoped (do not replace)

```ez
using std

do main() {
    // Has: std
}

do helper() {
    using arrays
    // Has: std (from file) + arrays (from function)
}

do another() {
    // Has: std only (arrays was function-scoped to helper)
}
```

### 3. Ambiguity Resolution

When multiple modules contain the same function name, **error on ambiguity**:

```ez
import @std
import @mymodule

using std       // Has println()
using mymodule  // Also has println()

do main() {
    println("test")         // ❌ ERROR: ambiguous function call
    std.println("test")     // ✅ Explicit prefix required
    mymodule.println("hi")  // ✅ Explicit prefix required
}
```

**Error message:**
```
error[E3006]: ambiguous function call
  --> main.ez:8:5
   |
8  |     println("test")
   |     ^^^^^^^ function 'println' found in multiple modules
   |
   = note: found in module 'std'
   = note: found in module 'mymodule'
   = help: use explicit module prefix: std.println() or mymodule.println()
```

### 4. Multiple Modules Per Statement

Both single and comma-separated syntax are allowed (programmer's choice):

```ez
// Single module per statement
using std
using arrays
using strings

// Multiple modules per statement
using std, arrays, strings

// Mixed (valid)
using std, arrays
using strings
```

**Implementation note**: Parser should handle comma-separated list in `using` statements.

### 5. Error Detection

All the following cases should produce clear error messages:

#### Error 1: Using before import
```ez
using std      // ❌ ERROR
import @std
```

**Error:**
```
error[E5004]: module must be imported before use
  --> main.ez:1:7
   |
1  | using std
   |       ^^^ module 'std' not yet imported
   |
   = help: move import statement before this using declaration
```

#### Error 2: Using non-imported module
```ez
import @std
using arrays   // ❌ ERROR: arrays not imported
```

**Error:**
```
error[E5001]: module not imported
  --> main.ez:2:7
   |
2  | using arrays
   |       ^^^^^^ module 'arrays' has not been imported
   |
   = help: add 'import @arrays' before this using declaration
```

#### Error 3: Function-scoped using of non-imported module
```ez
import @std
using std

do main() {
    using arrays  // ❌ ERROR: arrays not imported
}
```

**Error:**
```
error[E5001]: module not imported
  --> main.ez:5:11
   |
5  |     using arrays
   |           ^^^^^^ module 'arrays' has not been imported
   |
   = help: add 'import @arrays' at the file level
```

#### Error 4: Using in invalid scope
```ez
const MyStruct struct {
    using std  // ❌ ERROR
    name string
}
```

**Error:**
```
error[E1003]: invalid syntax
  --> main.ez:2:5
   |
2  |     using std
   |     ^^^^^ 'using' declarations not allowed inside struct definitions
   |
   = help: move 'using' declaration to file or function scope
```

#### Error 5: Using after other declarations
```ez
import @std

const MyStruct struct { }

using std  // ❌ ERROR: too late
```

**Error:**
```
error[E1003]: invalid placement of 'using' declaration
  --> main.ez:5:1
   |
5  | using std
   | ^^^^^ file-scoped 'using' must come before all declarations
   |
   = note: 'using' declarations must appear after imports but before any structs, constants, or functions
   = help: move this 'using' declaration to line 3 (after imports)
```

### 6. Keyword Choice

Use the `using` keyword for both file-scoped and function-scoped declarations:
- Placement determines scope (file-level vs function-level)
- Same keyword keeps the language consistent
- Can be aliased later via keyword aliasing feature (see issue #52)

```ez
using std      // File-scoped (outside functions)

do main() {
    using arrays  // Function-scoped (inside functions)
}
```

## Implementation Plan

### Phase 1: AST Changes

Add support for file-scoped `using` in the AST:

```go
type Program struct {
    Imports    []*ImportStatement
    Using      []*UsingStatement    // NEW: File-scoped using
    Statements []Statement
}

type UsingStatement struct {
    Token   Token
    Modules []string  // Support comma-separated: ["std", "arrays"]
}
```

### Phase 2: Parser Changes

Modify `pkg/parser/parser.go`:

1. **Parse file-scoped `using` after imports**:
   ```go
   func (p *Parser) ParseProgram() *Program {
       program := &Program{}
       
       // Parse imports
       for p.currentTokenMatches(IMPORT) {
           program.Imports = append(program.Imports, p.parseImportStatement())
       }
       
       // Parse file-scoped using statements
       for p.currentTokenMatches(USING) {
           program.Using = append(program.Using, p.parseUsingStatement())
       }
       
       // Parse rest of program
       for !p.currentTokenMatches(EOF) {
           stmt := p.parseStatement()
           if stmt != nil {
               // Error if we encounter 'using' here (too late)
               if _, ok := stmt.(*UsingStatement); ok {
                   p.addError("file-scoped 'using' must come before declarations")
               }
               program.Statements = append(program.Statements, stmt)
           }
       }
       
       return program
   }
   ```

2. **Parse comma-separated modules**:
   ```go
   func (p *Parser) parseUsingStatement() *UsingStatement {
       stmt := &UsingStatement{Token: p.currentToken}
       
       p.nextToken() // move past 'using'
       
       // First module
       stmt.Modules = append(stmt.Modules, p.currentToken.Literal)
       
       // Check for comma-separated modules
       for p.peekTokenMatches(COMMA) {
           p.nextToken() // move to comma
           p.nextToken() // move to next module
           stmt.Modules = append(stmt.Modules, p.currentToken.Literal)
       }
       
       return stmt
   }
   ```

3. **Validate using placement**:
   ```go
   func (p *Parser) validateUsingPlacement(using *UsingStatement) {
       // Check that all modules are imported
       for _, module := range using.Modules {
           if !p.isModuleImported(module) {
               p.addError(fmt.Sprintf("module '%s' not imported", module))
           }
       }
   }
   ```

### Phase 3: Semantic Analysis

Add module scope tracking during semantic analysis:

```go
type Analyzer struct {
    fileModules     []string  // Modules from file-scoped 'using'
    functionModules []string  // Modules from function-scoped 'using'
}

func (a *Analyzer) resolveFunctionCall(name string) {
    // Check which modules have this function
    modules := a.findModulesWithFunction(name)
    
    if len(modules) > 1 {
        // Error: ambiguous
        a.addError(E3006, fmt.Sprintf("function '%s' found in multiple modules", name))
    } else if len(modules) == 0 {
        // Error: not found
        a.addError(E3002, fmt.Sprintf("undefined function '%s'", name))
    }
}
```

### Phase 4: Error Code Addition

Add to `pkg/errors/codes.go`:

```go
// E3006: Ambiguous function call
E3006 = ErrorCode{"E3006", "ambiguous-function", "Ambiguous function call"}

// E5004: Module must be imported before use
E5004 = ErrorCode{"E5004", "module-not-imported-yet", "Module not imported yet"}
```

## Testing Plan

### Test Cases

#### 1. Basic File-Scoped Using
```ez
import @std
using std

do main() {
    println("Hello")  // ✓ std available
}
```

#### 2. Multiple File-Scoped Using
```ez
import @std
import @arrays
using std
using arrays

do main() {
    println("Hello")   // ✓ std
    push(arr, 1)       // ✓ arrays
}
```

#### 3. Comma-Separated Using
```ez
import @std
import @arrays
import @strings
using std, arrays, strings

do main() {
    println("Hello")      // ✓ std
    push(arr, 1)          // ✓ arrays
    concat("a", "b")      // ✓ strings
}
```

#### 4. Function-Scoped Adds to File-Scoped
```ez
import @std
import @arrays
using std

do main() {
    using arrays
    println("Hello")   // ✓ std (from file)
    push(arr, 1)       // ✓ arrays (from function)
}

do helper() {
    println("World")   // ✓ std (from file)
    push(arr, 1)       // ❌ ERROR: arrays not in scope
}
```

#### 5. Ambiguity Error
```ez
import @std
import @mymodule
using std, mymodule

do main() {
    println("test")         // ❌ ERROR: ambiguous
    std.println("test")     // ✓ explicit
}
```

#### 6. Using Before Import (Error)
```ez
using std      // ❌ ERROR
import @std
```

#### 7. Using Non-Imported Module (Error)
```ez
import @std
using arrays   // ❌ ERROR: not imported
```

#### 8. Using After Declarations (Error)
```ez
import @std
const X int = 10
using std      // ❌ ERROR: too late
```

#### 9. Using Inside Struct (Error)
```ez
const MyStruct struct {
    using std  // ❌ ERROR: invalid scope
}
```

#### 10. Mixed Styles
```ez
import @std
import @arrays
import @strings

using std, arrays  // Comma-separated
using strings      // Single

do main() {
    println("All three available")
}
```

## Documentation Requirements

### Language Reference Update

Add section on `using` declarations:

```markdown
## Using Declarations

The `using` keyword brings imported modules into scope, allowing you to call their functions without the module prefix.

### File-Scoped Using

Declare `using` at file level (after imports) to make modules available throughout the entire file:

```ez
import @std
import @arrays

using std      // Available in all functions
using arrays   // Available in all functions

do main() {
    println("Hello")    // std.println
    push(arr, 5)        // arrays.push
}

do helper() {
    println("World")    // std still available
}
```

### Function-Scoped Using

Declare `using` inside a function to add modules for that function only:

```ez
import @std
import @arrays

using std  // File-scoped

do main() {
    using arrays  // Add arrays for this function only
    println("Hello")  // std (from file)
    push(arr, 5)      // arrays (from function)
}

do helper() {
    println("World")  // std available
    push(arr, 5)      // ERROR: arrays not available here
}
```

### Multiple Modules

Use comma-separated syntax or multiple statements (programmer's choice):

```ez
// Style 1: One per line
using std
using arrays
using strings

// Style 2: Comma-separated
using std, arrays, strings

// Style 3: Mixed
using std, arrays
using strings
```

### Ambiguity

If multiple modules contain the same function, use explicit module prefix:

```ez
using std, mymodule

do main() {
    println("test")         // ERROR: ambiguous
    std.println("test")     // OK: explicit
    mymodule.println("hi")  // OK: explicit
}
```
```

### Style Guide Recommendation

```markdown
## Using Declaration Style

**File-scoped using:**
Use file-scoped `using` for modules needed throughout the file:
- `std` - if multiple functions use standard I/O
- Module-specific utilities used across functions

**Function-scoped using:**
Use function-scoped `using` for:
- Modules only needed in one function
- Avoiding namespace pollution

**Comma-separated syntax:**
Both styles are valid - choose what's most readable:
- Use comma-separated for related modules: `using std, io, fmt`
- Use separate lines for clearer diffs and version control
```

## Examples

### Before (Current Behavior)
```ez
import @std
import @arrays
import @strings

do main() {
    using std
    using arrays
    using strings
    
    println("Processing data...")
    push(myArr, 1)
    concat("hello", "world")
}

do helper() {
    using std
    using arrays
    
    println("Helper function")
    push(myArr, 2)
}

do another() {
    using std
    
    println("Another function")
}
```

### After (With File-Scoped Using)
```ez
import @std
import @arrays
import @strings

using std, arrays, strings  // File-scoped

do main() {
    println("Processing data...")
    push(myArr, 1)
    concat("hello", "world")
}

do helper() {
    println("Helper function")
    push(myArr, 2)
}

do another() {
    println("Another function")
}
```

## Compatibility

This feature is **backward compatible** - existing function-scoped `using` statements continue to work unchanged.

## Future Enhancements

- Wildcard imports: `using std.*` to import all from module
- Selective imports: `using std.{println, print}` to import specific functions
- Import aliases: `using std as io` for renaming modules
- Linter rule to detect redundant function-scoped `using` when file-scoped exists

## Acceptance Criteria

- [ ] File-scoped `using` declarations work after imports
- [ ] File-scoped `using` makes modules available in all functions
- [ ] Function-scoped `using` adds to (not replaces) file-scoped modules
- [ ] Comma-separated syntax works: `using std, arrays, strings`
- [ ] Error on using before import
- [ ] Error on using non-imported module
- [ ] Error on using after declarations
- [ ] Error on using in invalid scopes (structs, etc.)
- [ ] Error on ambiguous function calls with helpful message
- [ ] All test cases pass
- [ ] Documentation updated (language reference, style guide)
- [ ] Example programs demonstrating feature

--- a/examples/file_scoped_using.ez
+++ b/examples/file_scoped_using.ez
@@ -1,0 +1,59 @@
+import @std
+import @arrays
+
+/*
+ * EZ Language - File-Scoped Using Declarations
+ *
+ * File-scoped 'using' declarations make imported modules available
+ * throughout the entire file, eliminating the need to repeat 'using'
+ * statements in every function.
+ *
+ * Key features:
+ * - Must come after imports, before other declarations
+ * - Available in all functions in the file
+ * - Supports comma-separated syntax: using std, arrays
+ * - Function-scoped 'using' adds to (not replaces) file-scoped modules
+ */
+
+// File-scoped using - available in all functions
+using std, arrays
+
+do main() {
+    println("=== File-Scoped Using Examples ===")
+    println("")
+
+    example_basic()
+    example_function_scoped()
+}
+
+// Example 1: File-scoped using
+do example_basic() {
+    // std and arrays are already available via file-scoped using
+    println("1. File-scoped using:")
+
+    temp numbers [int] = {1, 2, 3}
+    println("  Original array:", numbers)
+
+    push(numbers, 4)
+    push(numbers, 5)
+    println("  After pushes:", numbers)
+
+    temp last int = pop(numbers)
+    println("  Popped:", last)
+    println("  Final array:", numbers)
+    println("")
+}
+
+// Example 2: Function-scoped using ADDS to file-scoped
+do example_function_scoped() {
+    // Already have std and arrays from file-scoped using
+    println("2. Function-scoped using adds to file-scoped:")
+
+    temp data [string] = {"hello", "world"}
+    println("  Data:", data)
+    println("  Length:", len(data))
+
+    // If we added 'using strings' here, we'd have std + arrays + strings
+    // But for this example, we just demonstrate that file-scoped is available
+    println("")
+}

--- a/pkg/ast/ast.go
+++ b/pkg/ast/ast.go
@@ -23,6 +23,7 @@ type Expression interface {
 
 // Program is the root node of every AST
 type Program struct {
+	FileUsing  []*UsingStatement // File-scoped using declarations
 	Statements []Statement
 }
 
@@ -386,10 +387,11 @@ type ImportStatement struct {
 func (i *ImportStatement) statementNode(){}
 func (i *ImportStatement) TokenLiteral() string { return i.Token.Literal }
 
-// UsingStatement represents using module
+// UsingStatement represents using module(s)
+// Supports both single (using std) and comma-separated (using std, arrays)
 type UsingStatement struct {
-	Token  Token
-	Module *Label
+	Token   Token
+	Modules []*Label // List of modules to bring into scope
 }
 
 func (u *UsingStatement) statementNode(){}

--- a/pkg/errors/codes.go
+++ b/pkg/errors/codes.go
@@ -40,6 +40,7 @@ var (
 	E3003 = ErrorCode{"E3003", "undefined-field", "field does not exist"}
 	E3004 = ErrorCode{"E3004", "undefined-module", "module not found"}
 	E3005 = ErrorCode{"E3005", "not-imported", "module not imported"}
+	E3006 = ErrorCode{"E3006", "ambiguous-function", "function found in multiple modules"}
 
 	// Runtime Errors (E4xxx)
 	E4001 = ErrorCode{"E4001", "division-by-zero", "cannot divide by zero"}
@@ -54,6 +55,7 @@ var (
 	E5001 = ErrorCode{"E5001", "circular-import", "circular import detected"}
 	E5002 = ErrorCode{"E5002", "file-not-found", "file not found"}
 	E5003 = ErrorCode{"E5003", "invalid-module", "invalid module"}
+	E5004 = ErrorCode{"E5004", "module-not-imported-yet", "module not imported yet"}
 )
 
 // Warning code definitions

--- a/pkg/interpreter/evaluator.go
+++ b/pkg/interpreter/evaluator.go
@@ -146,14 +146,16 @@ func Eval(node ast.Node, env *Environment) Object {
 		return NIL
 
 	case *ast.UsingStatement:
-		// Bring the module's functions into scope
-		alias := node.Module.Value
-		// Verify the module was imported
-		if _, ok := env.GetImport(alias); !ok {
-			return newErrorWithLocation("E3005", node.Token.Line, node.Token.Column,
-				"cannot use '%s': module not imported", alias)
+		// Bring the module(s) functions into scope (function-scoped using)
+		for _, module := range node.Modules {
+			alias := module.Value
+			// Verify the module was imported
+			if _, ok := env.GetImport(alias); !ok {
+				return newErrorWithLocation("E3005", node.Token.Line, node.Token.Column,
+					"cannot use '%s': module not imported", alias)
+			}
+			env.Use(alias)
 		}
-		env.Use(alias)
 		return NIL
 
 	// Expressions
@@ -243,7 +245,36 @@ func Eval(node ast.Node, env *Environment) Object {
 func evalProgram(program *ast.Program, env *Environment) Object {
 	var result Object
 
+	// First, process import statements
 	for _, stmt := range program.Statements {
+		if importStmt, ok := stmt.(*ast.ImportStatement); ok {
+			result = Eval(importStmt, env)
+			if isError(result) {
+				return result
+			}
+		}
+	}
+
+	// Then, process file-scoped using declarations
+	for _, usingStmt := range program.FileUsing {
+		for _, module := range usingStmt.Modules {
+			alias := module.Value
+			// Verify the module was imported
+			if _, ok := env.GetImport(alias); !ok {
+				return newErrorWithLocation("E3005", usingStmt.Token.Line, usingStmt.Token.Column,
+					"cannot use '%s': module not imported", alias)
+			}
+			env.Use(alias)
+		}
+	}
+
+	// Finally, process all other statements
+	for _, stmt := range program.Statements {
+		// Skip imports (already processed)
+		if _, ok := stmt.(*ast.ImportStatement); ok {
+			continue
+		}
+
 		result = Eval(stmt, env)
 
 		switch result := result.(type) {
@@ -720,13 +751,34 @@ func evalIdentifier(node *ast.Label, env *Environment) Object {
 	}
 
 	// Check if the function is available via "using" modules
+	// Detect ambiguity: if multiple modules have the same function, error
+	var foundModules []string
+	var foundBuiltin *Builtin
+
 	for _, alias := range env.GetUsing() {
 		if module, ok := env.GetImport(alias); ok {
 			fullName := module + "." + node.Value
 			if builtin, ok := builtins[fullName]; ok {
-				return builtin
+				foundModules = append(foundModules, module)
+				foundBuiltin = builtin
 			}
 		}
+	}
+
+	// Ambiguity check
+	if len(foundModules) > 1 {
+		err := newErrorWithLocation("E3006", node.Token.Line, node.Token.Column,
+			"function '%s' found in multiple modules", node.Value)
+		// Build helpful error message
+		moduleList := strings.Join(foundModules, ", ")
+		err.Help = fmt.Sprintf("use explicit module prefix: %s.%s()", foundModules[0], node.Value)
+		err.Message = fmt.Sprintf("function '%s' found in multiple modules: %s", node.Value, moduleList)
+		return err
+	}
+
+	// Found in exactly one module
+	if len(foundModules) == 1 {
+		return foundBuiltin
 	}
 
 	// Check global builtins (like len, typeof, etc.)

--- a/pkg/parser/parser.go
+++ b/pkg/parser/parser.go
@@ -351,9 +351,32 @@ func (p *Parser) curPrecedence() int {
 // ParseProgram parses the entire program
 func (p *Parser) ParseProgram() *Program {
 	program := &Program{}
+	program.FileUsing = []*UsingStatement{}
 	program.Statements = []Statement{}
 
+	seenOtherDeclaration := false
+
 	for !p.currentTokenMatches(EOF) {
+		// Track what we've seen for placement validation
+		if p.currentTokenMatches(USING) {
+			// File-scoped using: must come before other declarations
+			if seenOtherDeclaration {
+				p.addEZError(errors.E1003, "file-scoped 'using' must come before all declarations", p.currentToken)
+				p.nextToken()
+				continue
+			}
+
+			// Parse and add to FileUsing
+			usingStmt := p.parseUsingStatement()
+			if usingStmt != nil {
+				program.FileUsing = append(program.FileUsing, usingStmt)
+			}
+			p.nextToken()
+			continue
+		} else if !p.currentTokenMatches(EOF) && !p.currentTokenMatches(IMPORT) {
+			seenOtherDeclaration = true
+		}
+
 		stmt := p.parseStatement()
 		if stmt != nil {
 			program.Statements = append(program.Statements, stmt)
@@ -1273,12 +1296,23 @@ func (p *Parser) parseImportStatement() *ImportStatement {
 
 func (p *Parser) parseUsingStatement() *UsingStatement {
 	stmt := &UsingStatement{Token: p.currentToken}
+	stmt.Modules = []*Label{}
 
+	// Parse first module
 	if !p.expectPeek(IDENT) {
 		return nil
 	}
 
-	stmt.Module = &Label{Token: p.currentToken, Value: p.currentToken.Literal}
+	stmt.Modules = append(stmt.Modules, &Label{Token: p.currentToken, Value: p.currentToken.Literal})
+
+	// Check for comma-separated modules
+	for p.peekTokenMatches(COMMA) {
+		p.nextToken() // move to comma
+		if !p.expectPeek(IDENT) {
+			return nil
+		}
+		stmt.Modules = append(stmt.Modules, &Label{Token: p.currentToken, Value: p.currentToken.Literal})
+	}
 
 	return stmt
 }

